### PR TITLE
fix preserve-fds flag may cause runc hang

### PR DESCRIPTION
**Pull request by lifubang**
`--preserve-fds` may caluse `runc create` and `runc run` hang.
If preserve-fds > 3, and /dev/fd/5 is not exist, runc will hang.
Because at that time, there is a socket pair for `runc init`, 
one of it's fd is 5.
So, when we pass preserve-fds flag, we need to check whether these fds is exist.

```
root@test:/opt/ubuntu# ls /dev/fd/ -alh
dr-x------ 2 root root  0 2月  28 16:36 .
dr-xr-xr-x 9 root root  0 2月  28 16:36 ..
lrwx------ 1 root root 64 2月  28 16:36 0 -> /dev/pts/2
lrwx------ 1 root root 64 2月  28 16:36 1 -> /dev/pts/2
lrwx------ 1 root root 64 2月  28 16:36 2 -> /dev/pts/2
lr-x------ 1 root root 64 2月  28 16:36 3 -> /proc/24479/fd
root@test:/opt/ubuntu# docker-runc run -d --preserve-fds 3 test1
^C^C^C^C^C
root@test:/opt/ubuntu#
```

Signed-off-by: lifubang <lifubang@acmcoder.com>

### CODE REVIEW

File: utils_linux.go [hunk 1]
```diff
@@ -281,6 +281,11 @@ func (r *runner) run(config *specs.Process) (int, error) {
 	}
 	baseFd := 3 + len(process.ExtraFiles)
 	for i := baseFd; i < baseFd+r.preserveFDs; i++ {
+		_, err := os.Stat(fmt.Sprintf("/dev/fd/%d", i))
```
Comment (cyphar) [inline utils_linux.go]: Nit: `/proc/self/fd`.

File: utils_linux.go [hunk 2]
```diff
@@ -281,6 +281,11 @@ func (r *runner) run(config *specs.Process) (int, error) {
...
+		if err != nil {
+			r.destroy()
+			return -1, err
```
Comment (cyphar) [inline utils_linux.go]: It'd be nice if you could wrap the error to explain what's going on -- a user isn't going to understand what `stat /proc/self/fd/150: no such file or directory` means. Something like:

```
return -1, errors.Wrap(err, "check that all preserve-fds are present")
```

Comment (lifubang) [reply utils_linux.go]: +1 Thanks.

**Comment by cyphar** [comment]
LGTM.

**Comment by crosbymichael** [comment]
LGTM

**Pull request merged**